core/arch/arm/include/sm/optee_smc.h

```diff
@@ -93,6 +93,12 @@
 #define OPTEE_SMC_SHM_CACHED		1
 
 /*
+ * Privileged call flag. Used by supplicant. Asks OP-TEE to use reserved
+ * thread if all regular threads are used.
+ */
+#define OPTEE_SMC_FLAGS_PRIVILEGED	BIT(0)
```

Comment:
As I can see this, one extra privileged thread is sufficient. Can you give me example of calls where this will be not enough? 

We can use suspended thread's stack to register SHM, but there will be additional code to handle this specific case. I can do this if everyone is fine with this.

`OPTEE_MSG_ATTR_TYPE_NESTED` was needed because we need space to store list of pages. RPC command buffer have limited size (2 or 4 parameters, I don't quite remember) and it is allocated at OP-TEE side, so we can't reallocate it on kernel side. `OPTEE_MSG_ATTR_TYPE_NESTED` denoted PA of first page with actual list `struct optee_msg_param`. But, this again introduced new logic: code in `thread.c` needed to map additional pages, read list of parameters, construct new mobj... I was tired with debugging this, so I ditched it in favor of reusing existing calls that already registered shared memory. Code became much smaller and simpler. 

The existing approach is smallest in codebase and simplest of all variants I have tried. But, I can miss something, so I'm open to new ideas.